src/java.base/share/native/libjli/java.c

```diff
@@ -620,12 +618,33 @@ JavaMain(void* _args)
      * The main method is invoked here so that extraneous java stacks are not in
      * the application stack trace.
      */
-    if (!invokeStaticMainWithArgs(env, mainClass, mainArgs) &&
-        !invokeInstanceMainWithArgs(env, mainClass, mainArgs) &&
-        !invokeStaticMainWithoutArgs(env, mainClass) &&
-        !invokeInstanceMainWithoutArgs(env, mainClass)) {
-        ret = 1;
-        LEAVE();
+
+#define MAIN_WITH_ARGS 0
+#define MAIN_WITHOUT_ARGS 1
+#define MAIN_NONSTATIC 2
+
+    jclass helperClass = GetLauncherHelperClass(env);
+    jmethodID getMainType =
+        (*env)->GetStaticMethodID(env, helperClass, "getMainType", "()I");
+
+    int mainType = (*env)->CallStaticIntMethod(env, helperClass, getMainType);
+    int res = 0;
+    switch (mainType) {
+    case MAIN_WITH_ARGS:
+        res = invokeStaticMainWithArgs(env, mainClass, mainArgs);
+        break;
+    case MAIN_WITHOUT_ARGS:
+        res = invokeStaticMainWithoutArgs(env, mainClass);
+        break;
+    case MAIN_NONSTATIC:
+        res = invokeInstanceMainWithArgs(env, mainClass, mainArgs);
+        break;
+    case MAIN_NONSTATIC | MAIN_WITHOUT_ARGS:
+        res = invokeInstanceMainWithoutArgs(env, mainClass);
+        break;
+    }
```

Comment:
Only one of invokeXXXMainYYYArgs is called.   Looks like `CHECK_EXCEPTION_FAIL` and `CHECK_EXCEPTION_NULL_FAIL` are not needed and can use `CHECK_EXCEPTION_LEAVE` and `CHECK_EXCEPTION_NULL_LEAVE` instead?